File: app/models/discount_code.rb
class DiscountCode < ApplicationRecord
  belongs_to :user, optional: true
  belongs_to :product, optional: true
  belongs_to :category, optional: true
  has_many :discount_code_usages, dependent: :destroy
  has_many :users, through: :discount_code_usages

  enum discount_type: { percentage: 0, fixed_amount: 1 }
  enum status: { active: 0, inactive: 1 }

  validates :code, presence: true, uniqueness: true
  validates :discount_value, presence: true, numericality: { greater_than: 0 }
  validates :min_purchase, numericality: { greater_than_or_equal_to: 0 }, allow_nil: true
  validates :expires_at, presence: true, if: :active? # Ensure that expires_at is set if the discount code is active
  validate :expires_at_cannot_be_in_the_past, if: :active?

  def expires_at_cannot_be_in_the_past
    if expires_at.present? && expires_at < Time.current
      errors.add(:expires_at, "can't be in the past")
    end
  end

  def apply_discount(order_total)
    return order_total if inactive? || expired?

    case discount_type
    when "percentage"
      order_total - (order_total * (discount_value / 100))
    when "fixed_amount"
      [ order_total - discount_value, 0 ].max
    end
  end

  def expired?
    expires_at.present? && expires_at < Time.current
  end

  def usage_count
    discount_code_usages.count
  end

  def used_by?(user)
    discount_code_usages.exists?(user: user)
  end

  def mark_as_used(user)
    discount_code_usages.create(user: user) unless used_by?(user)
  end

  def self.active_codes_for_user(user)
    where(status: :active).where.not(id: user.discount_code_usages.select(:discount_code_id))
  end

  def self.expired_codes
    where("expires_at < ?", Time.current)
  end

  def self.inactive_codes
    where(status: :inactive)
  end

  def self.codes_for_product(product_id)
    where(product_id: product_id)
  end

  def self.codes_for_category(category_id)
    where(category_id: category_id)
  end
end

File: frontend/src/types/discountCode.ts
import { User } from "./auth";
import { Category } from "./category";
import { Product } from "./product";

/**
 * Discount code type enum
 */
export enum DiscountType {
  PERCENTAGE = "percentage",
  FIXED_AMOUNT = "fixed_amount",
}

/**
 * Discount code status enum
 */
export enum DiscountStatus {
  ACTIVE = "active",
  INACTIVE = "inactive",
}

/**
 * Discount code interface
 */
export interface DiscountCode {
  id: string;
  code: string;
  discountType: DiscountType;
  discountValue: number;
  minPurchase: number | null;
  status: DiscountStatus;
  expiresAt: string | null;
  expired: boolean;
  usageCount: number;
  userId: string | null;
  productId: string | null;
  categoryId: string | null;
  createdAt: string;
  updatedAt: string;
  user?: User;
  product?: Product;
  category?: Category;
}

/**
 * Discount code form data interface
 */
export interface DiscountCodeFormData {
  code: string;
  discountType: DiscountType;
  discountValue: number;
  minPurchase?: number | null;
  expiresAt?: string | null;
  status?: DiscountStatus;
  userId?: string | null;
  productId?: string | null;
  categoryId?: string | null;
}

/**
 * Discount code validation params interface
 */
export interface DiscountCodeValidationParams {
  code: string;
  amount?: number;
  productId?: string;
  categoryId?: string;
}

/**
 * Discount code validation result interface
 */
export interface DiscountCodeValidationResult {
  valid: boolean;
  discountCode?: DiscountCode;
  message?: string;
}

/**
 * Discount code application params interface
 */
export interface DiscountCodeApplicationParams {
  amount: number;
  markAsUsed?: boolean;
}

/**
 * Discount code application result interface
 */
export interface DiscountCodeApplicationResult {
  originalAmount: number;
  discountedAmount: number;
  discountAmount: number;
  discountCode: DiscountCode;
}
